### PR TITLE
Initial review of polymer-live-editor frame server and client code

### PR DESCRIPTION
**Pull request by eblaine**
PTAL

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+var privateKey = fs.readFileSync('key.pem').toString();
+var certificate = fs.readFileSync('cert.pem').toString();
+
+var credentials = {key: privateKey, cert: certificate};
+
+const app = express();
+
+var https = require('https').Server(credentials, app);
+
+
```
Comment (e111077) [inline app.js]: nit: spacing

File: app.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// [START app]
+const express = require('express');
+const fs = require('fs');
+
+var privateKey = fs.readFileSync('key.pem').toString();
+var certificate = fs.readFileSync('cert.pem').toString();
+
+var credentials = {key: privateKey, cert: certificate};
+
+const app = express();
```
Comment (e111077) [inline app.js]: If you're gonna have `const` and `var`s, stick with `const` and `let`. If you choose ES5, just use all `var`s

File: app.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+// [START app]
```
Comment (samuelli) [inline app.js]: Remove

File: app.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+});
+
+// Start the server
+const PORT = process.env.PORT || 8080;
+https.listen(PORT, () => {
+  console.log(`App listening on port ${PORT}`);
+  console.log('Press Ctrl+C to quit.');
+});
+// [END app]
```
Comment (samuelli) [inline app.js]: Remove

File: app.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+ */
+
+'use strict';
+
+// [START app]
+const express = require('express');
+const fs = require('fs');
+
+var privateKey = fs.readFileSync('key.pem').toString();
+var certificate = fs.readFileSync('cert.pem').toString();
```
Comment (samuelli) [inline app.js]: Can we document how these files are created? Are they required?

File: app.js (hunk 1 above)
Comment (samuelli) [inline app.js]: Remove multiple empty lines

File: app.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+const app = express();
+
+var https = require('https').Server(credentials, app);
+
+
+var io = require('socket.io')(https);
+
+app.use('/src', express.static('src'));
+app.use('/bower_components', express.static(__dirname + '/bower_components'));
+app.use('/node_modules', express.static(__dirname + '/node_modules'));
```
Comment (samuelli) [inline app.js]: Odd that 34 is different to 35-36, looks like they should all be the same.

File: app.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+app.get('/', (req, res) => {
+  // TODO (eblaine): load whitelisted domains from file once the editor is live
+  var options = {
+    headers: {
+      'Content-Security-Policy': "frame-ancestors 'self' https://2017-08-29-dot-frizzle-server.appspot.com https://*.google.com:*/"
```
Comment (samuelli) [inline app.js]: Why is google.com listed there?

File: app.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  };
+
+  socket.on('html-import-response', responseCallback);
+  socket.on('html-import-error', errorCallback);
+
+});
+
+app.get('*', (req, res) => {
+  res.status(404).send('Not found');
+})
```
Comment (samuelli) [inline app.js]: ;

File: app.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    if (filename !== errorData.filename) return;
+  
+    res.status(404).send(errorData.message);
+    socket.removeListener('html-import-error', errorCallback);
+    socket.removeListener('html-import-response', responseCallback);
+  };
+
+  socket.on('html-import-response', responseCallback);
+  socket.on('html-import-error', errorCallback);
+
```
Comment (samuelli) [inline app.js]: Remove empty lines before end of block

File: manifest.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (samuelli) [inline manifest.json]: Why is this necessary?

File: package.json [hunk 11]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "frizzle-frame",
+  "description": "Server to run user code typed in `polymer-live-editor` elements",
+  "main": "app.js",
+  "version": "0.0.1",
+  "private": true,
+  "license": "Apache-2.0",
```
Comment (samuelli) [inline package.json]: License file is missing

File: package.json [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  "author": "Google Inc.",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/PolymerLabs/polymer-live-editor-server.git"
+  },
+  "engines": {
+    "node": ">=4.3.2",
```
Comment (samuelli) [inline package.json]: I think this is incorrect since `=>` is used.

File: package.json [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    "npm": "^5.0.3"
+  },
+  "scripts": {
+    "start": "node app.js",
+    "lint": "samples lint",
```
Comment (samuelli) [inline package.json]: What is samples? Not referenced anywhere else.

Comment (samuelli) [reply package.json]: Can we add eslint?

File: package.json [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    "express": "4.15.3",
+    "google-cloud": "^0.56.0",
+    "https": "^1.0.0",
+    "prpl-server": "^0.8.0",
+    "socket.io": "^2.0.3"
+  },
+  "devDependencies": {
+    "@google-cloud/nodejs-repo-tools": "1.4.16"
+  },
+  "cloud-repo-tools": {
```
Comment (samuelli) [inline package.json]: Where is this used?

File: src/server-app/editor-server.html [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+-->
+<link rel="import" href="../../bower_components/polymer/polymer-element.html">
+
+<dom-module id="editor-server">
```
Comment (samuelli) [inline src/server-app/editor-server.html]: I'm confused about the structure of how all this fits together. Is there a design doc/explainer of how this all fits together?

cc @e111077

File: src/server-app/server-app.html [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+-->
+<link rel="import" href="../../bower_components/polymer/polymer-element.html">
+<link rel="import" href="./editor-server.html">
+
+<dom-module id="server-app">
```
Comment (samuelli) [inline src/server-app/server-app.html]: This appears to just wrap `editor-server`. Is there a reason they aren't the same element?

**Review by samuelli — CHANGES_REQUESTED** [review]
Just an initial pass, a little confused about how this fits together with the other pieces.

File: app.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+});
+
+app.get('*', (req, res) => {
+  res.status(404).send('Not found');
+})
+
+io.on('connection', function(socket){
+  socket.on('client-connection', function() {
```
Comment (e111077) [inline app.js]: security pow-wow with Peter:

At the point of the connection, the user should provide a secure cookie (feel free to use your favorite node cookie library e.g. cookie-parser is one i used to use a while ago).

Then in memory, associate this cookie with the socket session ID.

Then when they request an update, check if the cookie matches the socket ID that they are requesting before requesting the html content from the socket.

Then on socket disconnect, delete the cookie from memory

File: app.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+app.get('/update-demo/:socketId/:filename', (req, res) => {
+  var socketId = req.params.socketId;
+  var filename = req.params.filename;
+  var socket = io.sockets.connected[socketId];
+  if (!socket) {
+    return res.status(400).send('Demo does not exist or was not initialized');
+  }
+
+  socket.emit('html-import-request', filename);
```
Comment (e111077) [inline app.js]: before doing this, make sure the cookie provided is the same one associated with this socket id (see comment below)

File: bower.json [hunk 19]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "server",
+  "main": "index.html",
+  "dependencies": {
+    "polymer": "Polymer/polymer#^2.0.0",
+    "paper-button": "^2.0.0"
```
Comment (e111077) [inline bower.json]: probably eventually want more than just paper-button...

File: app.js (hunk 6 above)
Comment (e111077) [inline app.js]: That will also serve the node modules with like express etc. perhaps you should should move the bower_components and node_modules along with their bower.json and a new package.json into a sub directory like src/ and serve that?

e.g.
```js
app.use('/bower_components', express.static(__dirname + '/src/bower_components'));
app.use('/node_modules', express.static(__dirname + '/src/node_modules'));
```

and your file structure will look like
```
/
L package.json
  node_modules
  app.js
  src
  L package.json
    bower.json
    node_modules
    bower_components
```

File: app.js [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+  socket.on('html-import-response', responseCallback);
+  socket.on('html-import-error', errorCallback);
+
+});
+
+app.get('*', (req, res) => {
+  res.status(404).send('Not found');
+})
+
+io.on('connection', function(socket){
```
Comment (e111077) [inline app.js]: comment pls

File: src/server-app/editor-server.html [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+      static get is() { return 'editor-server'; }
+
+      constructor() {
+        super();
+        window.addEventListener('message', this._onMessage.bind(this));
+      }
+
+      ready() {
+        super.ready();
+        if (!window.parent) return;
```
Comment (e111077) [inline src/server-app/editor-server.html]: comment pls

File: src/server-app/editor-server.html [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+
+        this.socket = io();     
+        this.socket.emit('client-connection');
```
Comment (e111077) [inline src/server-app/editor-server.html]: This is the part where the server should check the cookie and associate cookie with socket id

**Review by e111077 — CHANGES_REQUESTED** [review]

File: app.js (hunk 5 above)
Comment (e111077) [reply app.js]: ++ good in readme also a general how-to in the readme would be appreciated

File: app.js (hunk 6 above)
Comment (e111077) [reply app.js]: I just saw the src app.use. By moving node_modules and bower_components into src, you shouldn't need the other two app.use calls

File: manifest.json (hunk 10 above)
Comment (e111077) [reply manifest.json]: I think this was made with polymer init and this might be vesitigial

File: package.json [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+    "test": "npm run system-test",
+    "e2e-test": "samples test deploy",
+    "postinstall": "bower install"
+  },
+  "dependencies": {
+    "bower": "^1.8.0",
+    "express": "4.15.3",
+    "google-cloud": "^0.56.0",
+    "https": "^1.0.0",
+    "prpl-server": "^0.8.0",
```
Comment (e111077) [inline package.json]: don't think you need prpl-server

File: package.json [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+    "pretest": "npm run lint",
+    "system-test": "samples test app",
+    "test": "npm run system-test",
+    "e2e-test": "samples test deploy",
+    "postinstall": "bower install"
+  },
+  "dependencies": {
+    "bower": "^1.8.0",
+    "express": "4.15.3",
+    "google-cloud": "^0.56.0",
```
Comment (e111077) [inline package.json]: Dont think this is necessary either since you are using gce and not app engine

File: package.json [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+    "https": "^1.0.0",
+    "prpl-server": "^0.8.0",
+    "socket.io": "^2.0.3"
+  },
+  "devDependencies": {
+    "@google-cloud/nodejs-repo-tools": "1.4.16"
```
Comment (e111077) [inline package.json]: dont think this is needed

File: polymer.json [hunk 26]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (e111077) [inline polymer.json]: please delete this file since we aren't polymer building

File: src/server-app/editor-server.html (hunk 15 above)
Comment (e111077) [reply src/server-app/editor-server.html]: 1. Developer creates an iframe with src pointing to this server
2. This server's CSP checks if it is an okay domain to embed the iframe
3. Server serves editor-server which communicates via messages through the iframe on what code to render
4. User specifies code to render
5. editor-server receives this message and creates a new iframe and inserts content
6. the new iframe may have html imports which it queries from server
7. server communicates via websockets with this element to determine whether or not the file is user-generated (not in bower_components / node_modules)
8. server serves the content and we don't have any 404s and everything is super duper happy 😀

File: src/server-app/editor-server.html [hunk 27]
```diff
@@ -0,0 +1,159 @@
...
+        }.bind(this));
+
+      }
+
+      _onMessage(event) {
+        if (!event || !event.data || !this._serverSocketId) return;  
+
+        this.htmlContents = event.data.codeObj;
+
+        let container = this.$.container;
```
Comment (e111077) [inline src/server-app/editor-server.html]: weird spacing around these lines

File: src/server-app/editor-server.html [hunk 28]
```diff
@@ -0,0 +1,159 @@
...
+        if (this.loadingFrame) {
```
Comment (e111077) [inline src/server-app/editor-server.html]: please make loadingFrame protected _loadingFrame

**Review by e111077 — CHANGES_REQUESTED** [review]

File: app.js (hunk 7 above)
Comment (eblaine) [reply app.js]: leftover from appengine. removing now.

File: app.js (hunk 2 above)
Comment (eblaine) [reply app.js]: @e111077 is there any reason not to use `const` and `let`? (in this file, that is)

Comment (e111077) [reply app.js]: I'd prefer you use `const` and `let` over `var`s. and the server runs on node, not a browser, so you're free to use the newest features

**Comment by eblaine** [comment]
PTAL - thanks!

File: README.md [hunk 29]
```diff
@@ -1,2 +1,74 @@
 # polymer-live-editor-server
-Server to run user code typed in `polymer-live-editor` elements
+Server and client code for previewing the results of
+user code typed in `polymer-live-editor` elements. This 
+server is aware of all Polymer elements.
```
Comment (samuelli) [inline README.md]: "all Polymer elements" sounds like all elements made with Polymer.

Maybe "all Polymer-team elements", or "elements in the PolymerElements org"

File: README.md [hunk 30]
```diff
@@ -1,2 +1,74 @@
...
+
```
Comment (samuelli) [inline README.md]: What about non Polymer-team elements?

File: README.md [hunk 31]
```diff
@@ -1,2 +1,74 @@
...
+the user's code. If the code contains HTML imports for files other
+than those found in `bower_components` or `node_modules`, the 
+server-side code (in `app.js`) will use the socket alluded to above
+to resolve those imports. 
+
+As an example, say a `polymer-live-editor` user is editing
+two files: `custom-element.html` and `index.html`. Here are the contents
+of `custom-element.html`:
+
+    <link rel="import"  href="/bower_components/polymer/polymer-element.html">
```
Comment (samuelli) [inline README.md]: Use triple backtick syntax to get highlighting
```
```html
```

File: README.md [hunk 32]
```diff
@@ -1,2 +1,74 @@
...
+
+    <script>
+      // Define the class for a new element called custom-element
+      class CustomElement extends Polymer.Element {
+        static get is() { return "custom-element"; }
+        constructor() {
+            super();
+            this.textContent = "I'm a custom-element.";
+          }
```
Comment (samuelli) [inline README.md]: Indentation

File: README.md (hunk 31 above)
Comment (samuelli) [inline README.md]: I think removing the absolute `/` is more typical. @e111077

File: README.md [hunk 33]
```diff
@@ -1,2 +1,74 @@
...
+    <custom-element></custom-element>
+
+In this case, the client code will use `document.write()` to write the 
+contents of `index.html`. During this process, the server will use a socket
+to request the contents of `custom-element.html` and return the
+user code in `custom-element.html`. If no such file exists either on the filesystem
+or in the user code, then the server returns a 404. 
+
+
+
```
Comment (samuelli) [inline README.md:89]: Missing documentation on the key files

File: app.js [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+let configuredSession = session({
+  secret: secret,
+  cookie: {secure: true},
+  resave: false,
+  saveUninitialized: true,
+  store: new MemoryStore({
+    checkPeriod: 86400000 // prune expired entries every 24h 
+  })
+});
+app.set('trust proxy', 1) // trust first proxy
```
Comment (samuelli) [inline app.js]: ;

File: package.json [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+    "type": "git",
+    "url": "https://github.com/PolymerLabs/polymer-live-editor-server.git"
+  },
+  "engines": {
+    "node": "=>4.3.2",
+    "npm": "^5.0.3"
+  },
+  "scripts": {
+    "start": "node app.js",
+    "postinstall": "cd src && npm install"
```
Comment (samuelli) [inline package.json:18]: I would rather only one top level package.json. bower & socket.io are included twice.
@e111077: thoughts?

File: package.json [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/PolymerLabs/polymer-live-editor-server.git"
+  },
+  "engines": {
+    "node": "=>4.3.2",
+    "npm": "^5.0.3"
+  },
+  "scripts": {
+    "start": "node app.js",
```
Comment (samuelli) [inline package.json:17]: Can we add an eslint?

https://github.com/GoogleChrome/rendertron/blob/324ac99732c943d7b16aead2fceaa8b31a458eaa/package.json#L11-L55

Also Travis for CI. These can be in a followup PR.

File: src/bower.json [hunk 37]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "server",
+  "main": "index.html",
+  "dependencies": {
+    "polymer": "Polymer/polymer#^2.0.0",
+    "paper-button": "^2.0.0"
```
Comment (samuelli) [inline src/bower.json]: How do all the components come in? Is this going to have all the elements eventually?

File: src/editor-server/editor-server.html [hunk 38]
```diff
@@ -0,0 +1,159 @@
+<!--
+/**
+ * Copyright 2017, Google, Inc.
```
Comment (samuelli) [inline src/editor-server/editor-server.html]: You can remove all the /* *

File: src/editor-server/editor-server.html [hunk 39]
```diff
@@ -0,0 +1,159 @@
...
+          } 
+
+          iframe.gray {
+            -webkit-filter: grayscale(100%);
+            z-index: 1;
+          }
+
+          iframe.fade {
+            opacity: 0;
+          }
```
Comment (samuelli) [inline src/editor-server/editor-server.html]: indentation of this whole style block is off

File: src/editor-server/editor-server.html [hunk 40]
```diff
@@ -0,0 +1,159 @@
...
+        this.fileContents = event.data.codeObj;
+        var container = this.$.container;
+        if (this._loadingFrame) {
+          this._loadingFrame.onload = null;
+          container.removeChild(this._loadingFrame);
+        }
+
+        this._loadingFrame = document.createElement('iframe');
+        this._loadingFrame.setAttribute('role', 'none');
+        this._loadingFrame.setAttribute('src', '/update-demo/' + this._serverSocketId + '/')
```
Comment (samuelli) [inline src/editor-server/editor-server.html]: ;

File: src/editor-server/editor-server.html [hunk 41]
```diff
@@ -0,0 +1,159 @@
...
+          iframe.fade {
+            opacity: 0;
+          }
+
+    </style>
+    
+    <div id="container"> 
+    </div>
+
+
```
Comment (samuelli) [inline src/editor-server/editor-server.html]: Remove extra padded lines.

File: src/bower.json [hunk 42]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "server",
+  "main": "index.html",
+  "dependencies": {
+    "polymer": "Polymer/polymer#^2.0.0",
```
Comment (samuelli) [inline src/bower.json:5]: Is this Polymer loaded via an iframe and hence doesn't need to match the polymer on the docs site?

If so, does that mean polymer is loaded twice for a docs site user?

File: src/editor-server/editor-server.html [hunk 43]
```diff
@@ -0,0 +1,159 @@
...
+          :host {
+            display: block;
+            height: 100%;
+          }
+
+          #container {
+            height: 100%;
+          }
+
+           iframe {
```
Comment (samuelli) [inline src/editor-server/editor-server.html]: indentation

**Review by samuelli — CHANGES_REQUESTED** [review]

File: src/bower.json (hunk 37 above)
Comment (eblaine) [reply src/bower.json]: Yeah, I was going to start with installing the iron-, paper-, app-, and gold-elements.

File: src/bower.json (hunk 42 above)
Comment (eblaine) [reply src/bower.json:5]: To the first question, yes. A user can type Polymer 2 or hybrid syntax into the editor, even though the docs site uses Polymer 1 at the moment. (This is the case because we're looking at revamping the Polymer 2 Quick Tour using this element.)

To the second, also yes - @e111077 is there a way to avoid this?

File: package.json (hunk 35 above)
Comment (e111077) [reply package.json:18]: That would be ideal, but how can you split a node_modules folder so you don't serve the server libraries alongside polymer jsmodules?

File: src/bower.json (hunk 42 above)
Comment (e111077) [reply src/bower.json:5]: Only solution I can imagine is to check if the correct version of polymer is outside of the iframe and have it passed in. I don't agree that this is necessary

Comment (samuelli) [reply src/bower.json:5]: Sorry, I'm not talking about the demo frame. I think that's actually a third instance of Polymer.
`<editor-server>` loads Polymer. All the elements in https://github.com/PolymerLabs/polymer-live-editor load Polymer too. Are any of them overlapped or are they all distinct?

File: app.js [hunk 44]
```diff
@@ -0,0 +1,122 @@
...
+app.set('trust proxy', 1); // trust first proxy 
+app.use(configuredSession);
+
+const https = require('https').Server(credentials, app);
+const io = require('socket.io')(https);
+let iosess = require('socket.io-express-session');
+io.use(iosess(configuredSession));
+
+app.use('/src', express.static(__dirname + '/src'));
+app.use('/bower_components', express.static(__dirname + '/src/bower_components'));
```
Comment (e111077) [inline app.js:50]: I don't think this and node_modules is necessary anymore if they are included in `/src`

File: app.js [hunk 45]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  }
+  res.sendFile(__dirname + '/src/index.html', options);
+});
+
+app.get('/update-demo/:socketID/:filename', (req, res) => {
+  const socketID = req.params.socketID;
+  
+  if (socketID !== req.session.socketID) {
+    return res.status(401).send('Invalid credentials.');
```
Comment (e111077) [inline app.js:66]: 👏👏👏👏👏

File: src/bower.json (hunk 42 above)
Comment (eblaine) [reply src/bower.json:5]: Thanks for the clarification. Here's a list of where I see Polymer being loaded:

`codemirror-textarea` loads Polymer because it's meant to be its own element that one could `bower install`.  (However, I'm now noticing that the `bower.json` in `codemirror-textarea` should be using hybrid syntax and it isn't.) 

Each `polymer-live-editor-file` imports Polymer. Then `polymer-live-editor` imports `polymer-live-editor-file`. The bower.json allows 1.9-2. 

Then `polymer-live-editor-server` loads Polymer 2.

Is there any way I can reduce the number of loads? It does seem like a lot of imports for one element. Thanks!

Comment (e111077) [reply src/bower.json:5]: Ah you are correct. The client loads polymer for the `live-editor` (shared with their instance of polymer if they are using it on their domain where they include this element e.g. polymer-project.org). Then this server loads an instance of Polymer in [`editor-server`](https://github.com/PolymerLabs/polymer-live-editor-server/blob/eb4b40fe43bd860c1acff6faa1dd73e4f0273893/src/editor-server/editor-server.html) in an iframe to this domain, and then the `editor-server` creates an iframe with the same domain that refreshes on edit and may load polymer if they use an html import to polymer.

I think the only avoidable load of polymer is in `editor-server`; if we are able to make it without using Polymer. Do you think it's worth it?

Additionally, do you think it would be worth it introducing a time-based debounce on recreating the iframe on edit?

**Review by e111077 — CHANGES_REQUESTED** [review]

File: src/bower.json (hunk 42 above)
Comment (e111077) [reply src/bower.json:5]: @eblaine clarification, each live editor file imports polymer, but that instance is shared across all files since they are all in the same project (thus the browser resolves the url).

File: src/editor-server/editor-server.html [hunk 46]
```diff
@@ -0,0 +1,156 @@
...
+          container.removeChild(this._loadingFrame);
+        }
+
+        this._loadingFrame = document.createElement('iframe');
+        this._loadingFrame.setAttribute('role', 'none');
+        this._loadingFrame.setAttribute('src', '/update-demo/' + this._serverSocketId + '/');
+        container.appendChild(this._loadingFrame);
+
+        try {
+          this._loadingFrame.contentDocument.open();
```
Comment (e111077) [inline src/editor-server/editor-server.html]: nit: `var contentDoc = this._loadingFrame.contentDocument;` used a bunch of times to be made a variable.

File: src/bower.json (hunk 42 above)
Comment (eblaine) [reply src/bower.json:5]: @e111077 i suppose there's no reason editor-server.html *needs* to know what Polymer is, since it basically exists to do socket.io stuff and call document.write(). However, this repo (`polymer-live-editor-server`) will still need Polymer as a dependency.

That said, I imagine most of the starter code blocks would HTML import `polymer-element` for learning purposes, so it might make sense for editor-server.html to load it anyway to prevent re-loading it all the time. What do you think?

File: app.js (hunk 44 above)
Comment (eblaine) [reply app.js:50]: It's currently necessary due to the base tag I'm adding to the innermost iframe. Without this, requests for bower_components in user code from the editor will search for /update-demo/:socketID/:filename/src/bower_components.

**Review by e111077 — CHANGES_REQUESTED** [review]

File: src/bower.json (hunk 42 above)
Comment (e111077) [reply src/bower.json:5]: So if you have an HTML import to something and then do an html import inside of an iframe with no `src` defined, does it actually transmit bytes, or is it already resolved? I believe that it still xmits bytes, but I'm not sure. Can you check in Chrome dev tools since you have an instance running?

Comment (eblaine) [reply src/bower.json:5]: Yes, you're right - looks like it still transmits bytes. So I suppose that means I should (1) keep polymer as a dependency in bower.json and (2) remove polymer from editor-server.html - is this correct?

Comment (samuelli) [reply src/bower.json:5]: @eblaine that sounds good to me!

File: app.js (hunk 44 above)
Comment (eblaine) [reply app.js:50]: however I was able to remove the node_modules one, because (for now) users shouldn't be importing anything from node_modules

**Comment by eblaine** [comment]
PTAL

File: README.md [hunk 47]
```diff
@@ -1,2 +1,90 @@
...
+the user's code. If the code contains HTML imports for files other
+than those found in `bower_components` or `node_modules`, the 
+server-side code (in `app.js`) will use the socket alluded to above
+to resolve those imports. 
+
+As an example, say a `polymer-live-editor` user is editing
+two files: `custom-element.html` and `index.html`. Here are the contents
+of `custom-element.html`:
+```html
+    <link rel="import"  href="/bower_components/polymer/polymer-element.html">
```
Comment (samuelli) [inline README.md]: Just a nit: you may want to unindent these:
https://github.com/PolymerLabs/polymer-live-editor-server/blob/b56e73b1773840d113366b3296b7092096a89cac/README.md

**Review by samuelli — APPROVED** [review]
LGTM!

**Review by e111077 — APPROVED** [review]
